### PR TITLE
(Validator) Add hidden `shred_storage` and `shred_storage_size` arguments for FIFO

### PR DESCRIPTION
**Pull request by yhchiang-sol**
#### Summary of Changes
This PR adds two hidden arguments to the validator that allow users to use RocksDB's FIFO compaction for storing shreds.

        --shred-storage <SHRED_STORAGE>
            EXPERIMENTAL: Controls how RocksDB compacts shreds.  *WARNING*: You will lose your ledger data
            when you switch between options. Possible values are: 'level': stores shreds using RocksDB's default (level)
            compaction. 'fifo': stores shreds under RocksDB's FIFO compaction. This option is more efficient on
            disk-write-bytes of the ledger store. [default: level]  [possible values: level, fifo]

        --shred-storage-size <SHRED_STORAGE_SIZE_BYTES>
            The shred storage size in bytes. The suggested value is 50% of your ledger storage size in bytes. [default:
            268435456000]

Dependency: #23236

#### Test Plan
Run validators with the following arguments:
* without using any rocksdb-compaction-style arguments -- expect to see the existing behavior:
```
/mnt/disks/ledger/mainnet $ ls
admin.rpc  calculate_accounts_hash_cache  contact-info.bin  genesis.bin  genesis.tar.bz2  ledger.lock  rocksdb  tmp-genesis
/mnt/disks/ledger/mainnet $ cd rocksdb/
/mnt/disks/ledger/mainnet/rocksdb $ grep FIFO OPTIONS-000077
/mnt/disks/ledger/mainnet/rocksdb $
```
* run with `--rocksdb-shred-compaction fifo` and `--rocksdb-fifo-shred-storage-size 536870912000`:
  - saw `rocksdb_fifo` directory is created as expected, and FIFO compaction is used as expected as well.
```
/mnt/disks/ledger/mainnet$ ls
admin.rpc  calculate_accounts_hash_cache  contact-info.bin  genesis.bin  genesis.tar.bz2  ledger.lock  rocksdb_fifo  tmp-genesis
/mnt/disks/ledger/mainnet$  cd rocksdb_fifo/
/mnt/disks/ledger/mainnet/rocksdb_fifo $ grep FIFO OPTIONS-000077
  compaction_style=kCompactionStyleFIFO
  compaction_style=kCompactionStyleFIFO
/mnt/disks/ledger/mainnet/rocksdb_fifo $
```
* run with `--rocksdb-shred-compaction level` and expect to see the same behavior as the first one (default).

### CODE REVIEW

File: validator/src/main.rs [hunk 1]
```diff
@@ -950,6 +954,32 @@ pub fn main() {
                 /* .default_value() intentionally not used here! */
                 .help("Keep this amount of shreds in root slots."),
         )
+        .arg(
+            Arg::with_name("shred_storage")
```
Comment (mvines) [inline validator/src/main.rs:958]: `—shred-storage` seems to not be quite what this option is about.  Perhaps `—rocksdb-shred-compaction [level|fifo]`?

File: validator/src/main.rs [hunk 2]
```diff
@@ -950,6 +954,32 @@ pub fn main() {
...
+                .default_value("rocksdb")
+                .help("EXPERIMENTAL: The underlying ledger storage for shreds. \
+                       *WARNING*: You will lose your ledger data when you switch between options. \
+                       Possible values are: \
+                       'rocksdb': stores shreds using RocksDB's default (level) compaction. \
+                       'rocksdb-fifo': stores shreds under RocksDB's FIFO compaction. \
+                           This option is more efficient on disk-write-bytes of the ledger store."),
+        )
+        .arg(
+            Arg::with_name("shred_storage_size")
```
Comment (mvines) [inline validator/src/main.rs:973]: looks like this arg is fifo specific?  if so it probably ought to include “fifo” or “rocksdb-fifo” in its name somewhere

File: validator/src/main.rs (hunk 1 above)
Comment (yhchiang-sol) [reply validator/src/main.rs:958]: Using `shred-storage` is to reserve the possibility that we have completely different storage for shreds in the future, but yep for now it's only switching between RocksDB's compaction styles.

Comment (mvines) [reply validator/src/main.rs:958]: Thanks, and yeah there's no need to prematurely generalize arguments.  Who knows what the future holds.

**Comment by yhchiang-sol** [comment]
Manually test by specifying the same ledger path but with different --rocksdb-shred-compaction values. Observe the ledger correctly keep both rocksdb and rocksdb-fifo:

```
/mnt/disks/ledger/mainnet $ ls -l
total 828
srw------- 1 yueh_hsuan_solana_com yueh_hsuan_solana_com      0 Feb 23 22:20 admin.rpc
drwxrwxr-x 2 yueh_hsuan_solana_com yueh_hsuan_solana_com  69632 Feb 23 23:01 calculate_accounts_hash_cache
-rw-rw-r-- 1 yueh_hsuan_solana_com yueh_hsuan_solana_com 606488 Feb 23 23:01 contact-info.bin
-rw-r--r-- 1 yueh_hsuan_solana_com yueh_hsuan_solana_com 132347 Mar 16  2020 genesis.bin
-rw-rw-r-- 1 yueh_hsuan_solana_com yueh_hsuan_solana_com  20144 Feb 23 22:01 genesis.tar.bz2
-rw-rw-r-- 1 yueh_hsuan_solana_com yueh_hsuan_solana_com      0 Feb 23 22:01 ledger.lock
drwxrwxr-x 2 yueh_hsuan_solana_com yueh_hsuan_solana_com   4096 Feb 23 22:19 rocksdb
drwxrwxr-x 2 yueh_hsuan_solana_com yueh_hsuan_solana_com   4096 Feb 23 23:01 rocksdb_fifo
drwxrwxr-x 2 yueh_hsuan_solana_com yueh_hsuan_solana_com   4096 Feb 23 22:01 tmp-genesis
```

And the resulting `rocksdb-fifo` also uses the correct compaction style:
```
/mnt/disks/ledger/mainnet/rocksdb_fifo $ grep FIFO OPTIONS-000077
  compaction_style=kCompactionStyleFIFO
  compaction_style=kCompactionStyleFIFO
/mnt/disks/ledger/mainnet/rocksdb_fifo $
```

and `rocksdb` also use the default level compaction as expected:
```
/mnt/disks/ledger/mainnet/rocksdb $ grep FIFO OPTIONS-000077
/mnt/disks/ledger/mainnet/rocksdb $
```

**Comment by yhchiang-sol** [comment]
Included test plan in the PR description:

#### Test Plan
Run validators with the following arguments:
* without using any rocksdb-compaction-style arguments -- expect to see the existing behavior:
```
/mnt/disks/ledger/mainnet $ ls
admin.rpc  calculate_accounts_hash_cache  contact-info.bin  genesis.bin  genesis.tar.bz2  ledger.lock  rocksdb  tmp-genesis
/mnt/disks/ledger/mainnet $ cd rocksdb/
/mnt/disks/ledger/mainnet/rocksdb $ grep FIFO OPTIONS-000077
/mnt/disks/ledger/mainnet/rocksdb $
```
* run with `--rocksdb-shred-compaction fifo` and `--rocksdb-fifo-shred-storage-size 536870912000`:
  - saw `rocksdb_fifo` directory is created as expected, and FIFO compaction is used as expected as well.
```
/mnt/disks/ledger/mainnet$ ls
admin.rpc  calculate_accounts_hash_cache  contact-info.bin  genesis.bin  genesis.tar.bz2  ledger.lock  rocksdb_fifo  tmp-genesis
/mnt/disks/ledger/mainnet$  cd rocksdb_fifo/
/mnt/disks/ledger/mainnet/rocksdb_fifo $ grep FIFO OPTIONS-000077
  compaction_style=kCompactionStyleFIFO
  compaction_style=kCompactionStyleFIFO
/mnt/disks/ledger/mainnet/rocksdb_fifo $
```
* run with `--rocksdb-shred-compaction level` and expect to see the same behavior as the first one (default).

File: ledger/src/blockstore_db.rs [hunk 3]
```diff
@@ -35,6 +35,8 @@ use {
     thiserror::Error,
 };
 
+pub const DEFAULT_ROCKS_FIFO_SHRED_STORAGE_SIZE_BYTES: u64 = 250 * 1024 * 1024 * 1024;
```
Comment (mvines) [inline ledger/src/blockstore_db.rs:38]: Can you please add a comment that provides some context regarding this magic number.

File: validator/src/main.rs [hunk 4]
```diff
@@ -2512,6 +2543,27 @@ pub fn main() {
...
 
+    validator_config.shred_storage_type = match matches.value_of("rocksdb_shred_compaction") {
+        None => ShredStorageType::default(),
+        Some(shred_compaction_string) => match shred_compaction_string {
+            "level" => ShredStorageType::RocksLevel,
+            "fifo" => {
+                let shred_storage_size = match matches.value_of("rocksdb_fifo_shred_storage_size") {
+                    Some(_) => value_t_or_exit!(matches, "rocksdb_fifo_shred_storage_size", u64),
+                    None => DEFAULT_ROCKS_FIFO_SHRED_STORAGE_SIZE_BYTES,
+                };
```
Comment (mvines) [inline validator/src/main.rs:2554]: The additional error handling here is unnecessary as clap.rs already takes care of it
```suggestion
                let shred_storage_size = value_t_or_exit!(matches, "rocksdb_fifo_shred_storage_size", u64);
```

**Review by mvines — APPROVED** [review]

**Pull request merged**